osu.Game.Rulesets.Osu/Scoring/OsuHealthProcessor.cs

```diff
@@ -0,0 +1,234 @@
+
+                if (!fail && recovery < hpRecoveryAvailable)
+                {
+                    fail = true;
+                    testDrop *= 0.96;
+                    hpMultiplierNormal *= 1.01;
+                    failReason = $"recovery too low ({recovery} < {hpRecoveryAvailable})";
+                }
+
+                if (fail)
+                {
+                    OnIterationFail?.Invoke($"FAILED drop {testDrop}: {failReason}");
+                    continue;
+                }
+
+                OnIterationSuccess?.Invoke($"PASSED drop {testDrop}");
+                return testDrop;
+            } while (true);
+
+            void reduceHp(double amount)
+            {
+                currentHpUncapped = Math.Max(0, currentHpUncapped - amount);
+                currentHp = Math.Max(0, currentHp - amount);
+            }
+
+            void increaseHp(HitObject hitObject)
+            {
+                double amount = healthIncreaseFor(hitObject, hitObject.CreateJudgement().MaxResult);
+                currentHpUncapped += amount;
+                currentHp = Math.Max(0, Math.Min(1, currentHp + amount));
+            }
+        }
+
+        protected override double GetHealthIncreaseFor(JudgementResult result) => healthIncreaseFor(result.HitObject, result.Type);
+
+        private double healthIncreaseFor(HitObject hitObject, HitResult result)
+        {
+            double increase;
+
+            switch (result)
+            {
+                case HitResult.SmallTickMiss:
+                    return IBeatmapDifficultyInfo.DifficultyRange(Beatmap.Difficulty.DrainRate, -0.02, -0.075, -0.14);
+
+                case HitResult.LargeTickMiss:
+                    return IBeatmapDifficultyInfo.DifficultyRange(Beatmap.Difficulty.DrainRate, -0.02, -0.075, -0.14);
+
+                case HitResult.Miss:
+                    return IBeatmapDifficultyInfo.DifficultyRange(Beatmap.Difficulty.DrainRate, -0.03, -0.125, -0.2);
+
+                case HitResult.SmallTickHit:
+                    // This result always comes from the slider tail, which is judged the same as a repeat.
+                    increase = 0.02;
+                    break;
+
+                case HitResult.LargeTickHit:
+                    // This result comes from either a slider tick or repeat.
+                    increase = hitObject is SliderTick ? 0.015 : 0.02;
+                    break;
+
+                case HitResult.Meh:
+                    increase = 0.002;
+                    break;
+
+                case HitResult.Ok:
+                    increase = 0.011;
+                    break;
+
+                case HitResult.Good:
+                    increase = 0.024;
+                    break;
+
+                case HitResult.Great:
+                    increase = 0.03;
+                    break;
+
+                case HitResult.Perfect:
+                    // 1.1 * Great. Unused.
```

Comment:
Bit weird that these are in the osu! implementation but will never be used by it. This and the `Good` one.

Reply:
I've removed these.